### DISCUSSION TITLE
Unicorn integration

### DISCUSSION THREAD

**Discussion by GranMinigun** [dosbox-staging/dosbox-staging#1110] [Ideas]
[Unicorn](https://www.unicorn-engine.org/) is a CPU emulator, basically QEMU's heart ripped out. Version 2.0 is a complete rewrite and currently WIP.

The proposal made by @dreamer is to integrate it as an alternative to dynamic cores, and if it'll work well - potentially replace them altogether.

**Comment by dreamer** [comment]
Yup, I think it can benefit our codebase greatly to have CPU emulator that's better designed and actually documented. But it's a long-term goal.

**Reply by johnnovak** [reply]
Something with a semi-accurate cycle count would be nice. Although probably the pipelining and instruction scheduling of Pentiums cannot be emulated in software as it would slow things down by at least two orders of magnitude.

**Comment by johnnovak** [comment]
Related:

Improve the emulation of speed-sensitive games by introducing CPU instruction latency tables
https://github.com/dosbox-staging/dosbox-staging/issues/2751

If Unicorn doesn't emulate instruction latencies, then that's not so great, after all.

**Reply by johnnovak** [reply]
Ok, so apparently QEMU and by extension Unicorn is "just" a dynamic recompiler, so there is zero attempt at emulating instruction latencies in a cycle-accurate manner. So I guess switching to Unicorn doesn't have that many benefits then, if our current dynamic recompiler JustWorks(tm)...

Looks like there have been attempts to turn QEMU into a cycle-accurate emulator, at least I found a couple of research papers, such as this one:
https://hal.science/hal-02192275/document

**Comment by FeralChild64** [comment]
QEMU CPU emulation is probably mainly relevant for Windows 9x (maybe even NT series). Also might be good for testing to confirm that some bug is a CPU emulation bug.
